tst/com/amazon/corretto/crypto/provider/test/EvpSignatureSpecificTest.java

```diff
@@ -98,6 +138,48 @@ private static void testKeyTypeMismatch(final String algorithm, final String bas
         }
     }
 
+    @ParameterizedTest
+    @MethodSource("ecdsaBadSignatures")
+    @SuppressWarnings("unchecked")
+    public void testBadEcdsaSignature(final String algorithm, final BigInteger r, final BigInteger s)
+            throws GeneralSecurityException {
+        final Signature verifier = Signature.getInstance(algorithm, NATIVE_PROVIDER);
+        // We always use the largest key to avoid hash truncation confusion
+        final ECPublicKey pubKey = (ECPublicKey) ECDSA_PAIR.getPublic();
+        final byte[] signature;
+        byte[] rArr = r.toByteArray();
+        if (rArr[0] == 0) { // Trim leading zero byte if present
+            rArr = Arrays.copyOfRange(rArr, 1, rArr.length);
+        }
+        byte[] sArr = s.toByteArray();
+        if (sArr[0] == 0) { // Trim leading zero byte if present
+            sArr = Arrays.copyOfRange(sArr, 1, sArr.length);
+        }
+
+        if (algorithm.endsWith("inP1363Format")) {
+            // Just zeros of the appropriate length
+            final int elementLength = (pubKey.getParams().getOrder().bitLength() + 7) / 8;
+            signature = new byte[elementLength * 2];
+            System.arraycopy(rArr, 0, signature, elementLength - rArr.length, rArr.length);
+            System.arraycopy(sArr, 0, signature, 2 * elementLength - sArr.length, sArr.length);
+        } else {
+            // It isn't worth building full ASN.1 logic here, so we cover the easy cases of ones and zeros
+            // and leave the complicated cases of the full orders to the P1363 tests
+            // (which cover the same implementations) under the hood
+            assumeTrue(rArr.length == 1 && sArr.length == 1, "This test only supports {0,1} for non-P1363 sigs.");
+            signature = new byte[]{0x30, 0x06, 0x02, 0x01, 0x00, 0x02, 0x01, 0x00};
+            signature[4] = rArr[0];
+            signature[7] = sArr[0];
+
+        }
+        verifier.initVerify(pubKey);
+        try {
+            assertFalse(verifier.verify(signature));
+        } catch (final SignatureException ex) {
+            // We also allow a SignatureException.
```

Comment:
for an invalid signature (as opposed to a mismatched but otherwise valid signature), shouldn't we always be throwing SignatureException? from [oracle's javadoc](https://docs.oracle.com/javase/7/docs/api/java/security/Signature.html#verify(byte[])):

> Returns:
>     true if the signature was verified, false if not.
> Throws:
>     SignatureException - if this signature object is not initialized properly, the passed-in signature is improperly encoded or of the wrong type, if this signature algorithm is unable to process the input data provided, etc.

Reply:
The exact lines of what is an invalid signature are unclear and not well specified. In fact you can see that [JDK15+ returns false](https://github.com/corretto/corretto-17/commit/2d4103a3d929e05edca98e7703e0869077966be7#diff-af656439ff8511c928a2107a4806c6b669c8851db55d30d9092cdd96db84be56R227) rather than throws an exception in this case. All that matters is that we reject it. How is far less important.

Reply:
I see, thanks for the context. Because the JCA interfaces go to the trouble of stipulating different failure modes (throwing vs. returning), I would have expected that they precisely specify which error/failure cases correspond to throwing vs. returning. I suppose it's reasonable to rely on a reference implementation absent something more formal.